src/main/java/net/minecraftforge/common/ForgeTagHandler.java

```diff
@@ -246,166 +224,44 @@ public static void setCustomTagTypes(Set<ResourceLocation> customTagTypes)
-        if (!customTagTypes.isEmpty())
-        {
-            LOGGER.debug("Populated the TagCollectionManager with {} extra types", customTagTypes.size());
-        }
-    }
-
-    /**
-     * Updates the custom tag types' tags from reloading via NetworkTagManager
-     *
-     * @apiNote Internal: For use by NetworkTagManager
-     */
-    public static void updateCustomTagTypes(List<TagCollectionReaderInfo> tagCollectionReaders)
-    {
-        ImmutableMap.Builder<ResourceLocation, TagCollection<?>> builder = ImmutableMap.builder();
-        for (TagCollectionReaderInfo info : tagCollectionReaders)
-        {
-            builder.put(info.tagType, info.reader.build(info.tagBuilders));
-        }
-        customTagTypes = builder.build();
-    }
-
-    /**
-     * Updates the custom tag types' tags from packet
+     * Creates the completable future in a wrapper object to be loaded by the TagManager
+     * Used for tags of non-vanilla registries
      *
      * @apiNote Internal
      */
-    public static void updateCustomTagTypes(SyncCustomTagTypes packet)
+    @SuppressWarnings("unchecked")
+    public static <T> TagManager.LoaderInfo<T> makeCustomTagTypeLoader(StaticTagHelper<T> helper, ResourceManager manager, Executor executor, Map<ResourceLocation, TagLoader<?>> readers)
     {
-        customTagTypes = packet.getCustomTagTypes();
-        reinjectOptionalTagsCustomTypes();
+        return new TagManager.LoaderInfo<>(helper, CompletableFuture.supplyAsync(() -> ((TagLoader<T>) readers.get(helper.getKey().location())).loadAndBuild(manager), executor));
     }
 
     /**
-     * Gets the completable future containing the reload results for all custom tag types.
-     *
-     * @apiNote Internal: For use by NetworkTagManager
-     */
-    public static CompletableFuture<List<TagCollectionReaderInfo>> getCustomTagTypeReloadResults(ResourceManager resourceManager, Executor backgroundExecutor, Map<ResourceLocation, TagLoader<?>> readers)
-    {
-        CompletableFuture<List<TagCollectionReaderInfo>> customResults = CompletableFuture.completedFuture(new ArrayList<>());
-        for (Entry<ResourceLocation, TagLoader<?>> entry : readers.entrySet())
-        {
-            CompletableFuture<Map<ResourceLocation, Tag.Builder>> collectionFuture =
-                    CompletableFuture.supplyAsync(() -> entry.getValue().load(resourceManager), backgroundExecutor);
-            customResults = customResults.thenCombine(collectionFuture, (results, result) -> {
-                results.add(new TagCollectionReaderInfo(entry.getKey(), entry.getValue(), result));
-                return results;
-            });
-        }
-        return customResults;
-    }
-
-    /**
-     * Add all the missing optional tags back into the custom tag types tag collections
+     * Creates a map for serialization, copied from {@link TagCollection#serializeToNetwork} but adapted it since it requires
+     * a normal registry.
```

Comment:
I am not sure which is better/less maintenance, but in theory this could make use of `TagCollection#serializeToNetwork` by making use of the backing vanilla registries all the custom types have and passing it to that. See `ForgeRegistyTagProvider#wrapRegistry`.